Table of content:

1. Variables
2. Functions
3. Strings
4. Lists
5. Dictionary
6. Tuples
7. Booleans
8. Functions again
9. Loops : for
10. If statement
11. While statement
12. Libraries import
13. Numpy
14. Pandas

## Python Tutorial


Welcome to this Python tutorial. The Jupyter Notebook interface is composed of several cells, in which you can write some code. When you run a cell, the only lines executed are those in the cell. However, they are linked by a single kernel : it means that you can define a variable in a cell and use it in an other.

## 1) Variables

To run a cell, select it and press "execute", or shift + enter. Try to run the following cell :

In [1]:
print("Congratulations")

Congratulations


If your code includes some functions which should display some informations, this information will appear under the box, like with "Congratulations" in the previous example.

#### &#x270d; 1.1) Display the number 13 and the message "Python is great"

In [7]:
a = 13
print(a)
print(" Python is Great")

13
 Python is Great


#### &#x270d; 1.2) Create the variables $a = 5$, $b = 5.2$ and $c = 3 + 2j$

In [8]:
a = 5
b = 5.2
c = 3 + 2j

#### &#x270d; 1.3) Display $a$, $b$ and $c$ with the print function.  
They can be displayed with only one call by separating them with commas in the brackets.

In [9]:
print(a,b,c)

5 5.2 (3+2j)


#### &#x270d; 1.4) Display $a + b$ and $b * c$ without saving the product in a variable.

In [16]:
print(a + b )
print(b * c)

10.2
(15.600000000000001+10.4j)


#### &#x270d; 1.5) Check the type of $a$, $b$, $c$, then display and show the type of :  
$a + b$  
$a * b$  
$b - c$  
$a + b + c$  
$a * c$  
$c / a$  

In [22]:
print(type(a))
print(type(b))
print(type(c))
print(type(a + b))
print(type(a * b))
print(type(b + c))
print(type(a + b + c))
print(type(a * c))
print(type(c/a))

<class 'int'>
<class 'float'>
<class 'complex'>
<class 'float'>
<class 'float'>
<class 'complex'>
<class 'complex'>
<class 'complex'>
<class 'complex'>


## 2)Functions

#### &#x270d; 2.1) Save round(b) in a variable e, then display the type of :  
$b$  
$e$  
round

In [2]:
b = 10.44555
e = round(b)
print(type(b))
print(type(e))

<class 'float'>
<class 'int'>


#### &#x270d; 2.2) Create a function which takes as argument a value $x$, and returns $x + x^2$ (show how your function works with example)

In [12]:
# x = 20
# val = (x + (x * x))
# print(val)
def printValues():
    l = list()
    for x in range(1,21):
        l.append(x + (x*x))
print(l)
		
printValues()


[2, 6, 12, 20, 30, 42, 56, 72, 90, 110, 132, 156, 182, 210, 240, 272, 306, 342, 380, 420]


#### &#x270d; 2.3) Create a function which takes as argument a value $x$, prints $x$, $2x$, $3x$, and returns $x - 1$. (show how your function works with example)

In [24]:
def f(x):
       return x

def g(x):
        return x * 2
def h(x):
        return x ** 3

print(f(2))
print(g(2))
print(h(2))


2
4
8


#### &#x270d; 2.4) Write down a function that takes $x$ and $y$ as the arguments, and returns a rounded value of $x$ multiplied by the value of $y$. (show how your function works with example)

In [29]:
def h(x,y):
        return (round(x) * y)

h(3.5452,4)

16

## 3) Strings

#### &#x270d; 3.1) Create a string called birthday containing your birthday.

In [38]:
birthday = ('2000')
print("my birthday is"  + birthday)

my birthday is2000


#### &#x270d; 3.2) Using a and b, create a string "Hello my dear" and print it.

In [38]:
a = ('hello')
b = ('my dear')
print(a + b)

hellomy dear


#### &#x270d; 3.3) Print the number of letters of a and b using the function len(string)

In [39]:
len(a + b)

12

A string can be included in the middle of another one using `%s`

Example: 

`animal = "beef"`

`action = "love"`

`dream = "i would %s to eat a %s" %(action,animal)`

`--> dream = "i would love to eat a beef"`

#### &#x270d; 3.4) Create the string "I am 30 years old and I live in Bogota" using the following strings :

In [40]:
age = (30)
town = "bogota"
profile = "I am %s years old and i live in %s"%(age,town)
print(profile)

I am 30 years old and i live in bogota


## 4) Lists

To create a list, you need to use `[ ]`. If you're using a mac, the command is alt + shift + `(` or `)` .


#### &#x270d; 4.1) Create a list containing every letters from A to H, called letters.

In [45]:
letters = ["A", "B", "C", "D", "E", "F", "G", "H"]
print(letters)

['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H']


#### &#x270d; 4.2) Print the letter E using its index in the list letters.

Remember : index of the first element is $0$.

In [46]:
print(letters[4])

E


#### &#x270d; 4.3) Replace the letter G by Y in letters.

In [47]:
letters[6] = "Y"
print(letters)

['A', 'B', 'C', 'D', 'E', 'F', 'Y', 'H']


#### &#x270d; 4.4) Print the letter D in two different ways : once with its index, one with its index starting from the last element (negative index).

In [48]:
print(letters[3])
print(letters[-5])

D
D


#### &#x270d; 4.5) Add the letter P to the list letters, using the function letters.append()

In [49]:
letters.append('P')
letters

['A', 'B', 'C', 'D', 'E', 'F', 'Y', 'H', 'P']

#### &#x270d; 4.6) Create a sublist sublet = ["A","B","C"] from the list letters.

Remember : list[i:j] creates a sublist from list starting from the i th element to the j-1 th element.

In [50]:
subset = letters[:3]
subset

['A', 'B', 'C']

#### &#x270d; 4.7) Concatenate letters and sublet into letters.

In [51]:
conc = subset + letters
print(conc)

['A', 'B', 'C', 'A', 'B', 'C', 'D', 'E', 'F', 'Y', 'H', 'P']


#### &#x270d; 4.8)  Erase every occurence of C in letters.

Remember : use list.remove(string) to delete every occurence of string in list.

In [61]:
conc = subset + letters
print(conc)
rem_char = "C"
new_conc = conc.remove(rem_char)
print(conc)


['A', 'B', 'C', 'A', 'B', 'C', 'D', 'E', 'F', 'Y', 'H', 'P']
['A', 'B', 'A', 'B', 'C', 'D', 'E', 'F', 'Y', 'H', 'P']


#### &#x270d; 4.9) Erase the second occurence of A in letters.
Remember : use list.pop(index) to delete the index th element.

In [69]:
conc.pop(0)
print(conc)

['A', 'B', 'C', 'D', 'E', 'F', 'Y', 'H', 'P']


## 5) Dictionary

Dictionaries are similiar to lists, but instead of accessing an element through its index, we can use some keywords, which can be of any type.
When you create it, you indicate the keyword and then the associated variables.
Try to print every variable of the following dict :

In [70]:
dict = {'x' : 3, 'y' : True, 'z' : [False, 89], 5 : "banana",  True : False }

#### &#x270d; 5.1) A list of the keywords of dict can be accessed using the function `dict.keys()`, use it.

In [71]:
dict.keys()

dict_keys(['x', 'y', 'z', 5, True])

#### &#x270d; 5.2) Add a new value False to dict, indexed by the keyword Banana.


To add a new keyword, give the associated value as if the keyword already existed. Example:  

`dict = {}`

`dict[3] = "hi"`

`--> dict = {3 : "hi"}`

In [72]:
dict = {}
dict[3] = 'banana'
dict

{3: 'banana'}

#### &#x270d; 5.3) Delete the key `y` from dict.

You can use `dict.pop(key)` to do that.

In [73]:
dict.pop(3)

'banana'

#### &#x270d; 5.4) Try to delete the key `g`, which is not in dict.

In [182]:
dict.del('g')

SyntaxError: invalid syntax (<ipython-input-182-5c4707f94ab7>, line 1)

As you can see, instead of just ignoring the command, an error is returned. If you do not want the program to be stopped if the value is not existing, you can add another action to be performed if the key does not exist. It must be written as second parameter.

Addding None as second parameter prevents the program from being stopped by an unexisting key.

#### &#x270d; 5.5) Check this with the previous example.

In [184]:
del('g', None='True')

SyntaxError: invalid syntax (<ipython-input-184-e8a6b18a6de4>, line 1)

## 6) Tuples
Tuples are very similar to lists, except their shape cannot be changed.

#### &#x270d; 6.1) Create a tuple containing the integers from 5 to 9 and a tuple containing the integers from 1 to 4. Use ( ) to create them.

In [79]:
int = (5,6,7,8,9)
print(int)
integer = (1,2,3,4)
integer

(5, 6, 7, 8, 9)


(1, 2, 3, 4)

#### &#x270d; 6.2) Try to sum up these tuples.

In [80]:
sum = int + integer
sum

(5, 6, 7, 8, 9, 1, 2, 3, 4)

#### &#x270d; 6.3) Print the 5th element of the sum, using two different ways.

In [81]:
print(sum[5])

sum[-4]

1


1

## 7) Booleans


#### &#x270d; 7.1) Create two booleans `a = True` and `b = False`, and check the result of every combination of `a` and `b` with the logical operators `and` / `or`.

In [90]:
colors = ["blue", "green"]

if "red"  in colors:
    print(blue & green)

#### &#x270d; 7.2) To invert a boolean, use the function not. Check the result of "not a or b"

In [87]:
colors = ["blue", "green"]

if "red" not in colors:
    print(True)

True


#### &#x270d; 7.3) Check how to use the different logical operators :  
equals ==  
less than <  
more than >    
less or equal <=
more or equal >=



In [104]:
print(10 > 9)
print(10 == 9)
print(10 < 9)
print(10 >= 9)
print(10 <= 9)

True
False
False
True
False


#### &#x270d; 7.4) Create a the boolean which indicates if 3 in in the list "list" or not.

Use the function in :  `a in list = True` if a is in list and `False` otherwise.

In [96]:
# list = ['a', 'b', 'c', 'd']
# a in list 
list = (1,2,3,4,5,6,7,8)
if 3  in list:
    print(True)
else:
        print(False)


True


## 8) Functions again


Once created, a function is used only if it is called.

Here, the function square returns the squared value of the input.

#### &#x270d; 8.1) Try to print the squared value of 5 using the square function below.

In [97]:
def square(n):
    result = n**2
    return(result)

    
print(square(5))


25


The "return" word is very important when you define a function, for two reasons. First, it allows to tell what will be the output of the function. Second, this word means that we are at the end of the function, so the next lines do not need to be read.

#### &#x270d; 8.2) Create a function `square2(n)` identical to `square(n)` without the `return(result)` line and see what happens if you try to compute `square2(5)`.

In [109]:
def square2(n):
    result = n**2
    

    
print(square(5))

25


#### &#x270d; 8.3) Create a function `square3(n)` identical to `square(n)` with one more line `print("hi")` after `return(result)` and see what happens if you try to compute `square3(5)`.

In [98]:
def square3(n):
    result = n**2
    return(result)
print("hi")

    
print(square3(5))

hi
25


#### &#x270d; 8.4) Create a function `firstandlast(l)` which returns a tuple containing the first and the last elements of `l`, a list input.
Try it on the list test.

In [111]:
def printValues():
    l = list()
    for i in range(1,21):
        l.append(i**2)
    print(l[:5])
    print(l[-5:])

printValues()


TypeError: 'tuple' object is not callable

Once a function is created, it can be called even in the definition of an other function.

#### &#x270d; 8.5) Create a function `sumoffandl` which takes as input a list, and returns the sum of its first and last terms. Use the `firstandlast(n)` function.

## 9) Loops : `for`


Run the following code and watch the syntaxe of the For loop function

In [112]:
my_list = [2,5,4,2,8,20]
for x in my_list:
    print(x)

2
5
4
2
8
20


#### &#x270d; 9.1) Now try your own loop : compute the sum of the terms of the list `my_list` with a for loop.

In [118]:
list = [2,5,4,2,8,20]
for x in list:
    print(x + x)

4
10
8
4
16
40


When we want to iterate on some integers, we use the range function : 

`for k in range(i,j)`

 means that `k` will take the value of every integer between `i` and `j-1`

#### &#x270d; 9.2) Compute the product of integers from 37 to 54 using a for loop and the range function.

In [113]:
for k in range(37, 54):
    print(k)

37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53


#### &#x270d; 9.3) Create a function which takes a list as input and returns the same list but elements are inverted (the first element becomes the last, etc)

In [115]:
def reverse(s): 
  str = "" 
  for i in s: 
    str = i + str
  return str
  
s = "The String To be reversed"
  
print ("The original string  is : ",end="") 
print (s) 
  
print ("The reversed string(using loops) is : ",end="") 
print (reverse(s))

The original string  is : The String To be reversed
The reversed string(using loops) is : desrever eb oT gnirtS ehT


## 10) `If` statement

#### &#x270d; 10.1) Test the function test with -5, 1, -1.7, 8, and interpret.

In [116]:
def test(x):
    if x**2 > -5 : #true/false
        print("yes")
    else :
        print("no")



#### &#x270d; 10.2) Create a function which takes a list as input and returns a boolean, set to `True` if the first element is a string, an `False` otherwise.

#### &#x270d; 10.3) Create a function which takes a list as input and returns a boolean, set to `True` if there is a string in the list, and `False` otherwise.

Hint : combine `for` and `if`.

## 11) `While` statement

#### &#x270d; 11.1) Run the following cell and interpret.

In [117]:
x = 16
while x > 3 :
    print(x)
    x = x / 2

16
8.0
4.0


#### &#x270d; 11.2) Create a function which takes a list, and returns false if there is no string inside, and the index of the first string otherwise, using a while loop.

#### &#x270d; 11.3) Print the first integer whose squared value is more than 7533493

## 12) Libraries import

### &#x270d; 12.1) Import the library numpy, with the shortened name np.

Remember : `import XXXXXX as YY` to import the module `XXXXXX` with the shortened name `YY`.

In [121]:
import numpy as np

Once a library is imported, all its functions are stored in the kernel and can be accessed.

#### &#x270d; 12.2) Print the value of pi using the function pi in numpy.

Remember : To use a function of a library imported as YY, use YY.function()

In [124]:
pi = np.pi
pi

3.141592653589793

## 13) Numpy

#### &#x270d; 13.1) Create an array containing the integers from 0 to 9

In [125]:
arr = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

print(arr)

[0 1 2 3 4 5 6 7 8 9]


#### &#x270d; 13.2) Create a 10x10 matrix name M, use arange and reshape

In [126]:
M = np.random.random((10,10))
print(M)

[[0.5757543  0.55005687 0.93339868 0.68006972 0.12863182 0.97887172
  0.90261551 0.6205993  0.64428114 0.01665926]
 [0.48120746 0.47431547 0.78824216 0.06918137 0.76707229 0.74062004
  0.79206006 0.98367348 0.08615552 0.93554057]
 [0.02469638 0.20801288 0.07322029 0.18855577 0.7705769  0.16038994
  0.94628664 0.53226626 0.16277958 0.7380185 ]
 [0.54864661 0.83269022 0.82129033 0.60682497 0.38078064 0.29677486
  0.33822887 0.65514967 0.20905579 0.51164197]
 [0.57147006 0.21104274 0.25661576 0.57388259 0.12723839 0.8858369
  0.17270516 0.65275381 0.22544264 0.14344637]
 [0.11561998 0.79368976 0.79521341 0.59286152 0.63242322 0.43717404
  0.47921474 0.31116176 0.34246043 0.17197375]
 [0.25175261 0.85484999 0.79146749 0.68943229 0.53978021 0.86348877
  0.52987913 0.51477118 0.95504903 0.70180609]
 [0.33220775 0.5363314  0.70013831 0.17385229 0.65000561 0.29543099
  0.86697894 0.25624995 0.07197663 0.76416896]
 [0.00447787 0.6084102  0.31898126 0.41037921 0.13318952 0.31646644
  0.28876778 

#### &#x270d; 13.4) Compute the sum of the lists X and Y without using the function for.

In [127]:
import numpy as np

X = np.array([1, 2, 3])

Y = np.array([4, 5, 6])

arr = np.concatenate((X, Y))

print(arr)

[1 2 3 4 5 6]


#### &#x270d; 13.5) Compute the scalar product of X and Y.

Hint : the numpy function is `dot`.

A square matrix, in which all diagonal elements are equal to same scalar and all other elements are zero, is called a scalar matrix. Or. A diagonal matrix, in which all diagonal elements are equal to same scalar, is called a scalar matrix. A.

In [129]:
np.dot(X,Y)

32

## 14) Pandas

In Pandas, we work with dataframes. Let's create one to see how it works.

#### &#x270d; 14.1) Import the pandas module, with pd as shortened name, and the random module. 

In [157]:
import pandas as pd
import numpy as np

#### &#x270d; 14.2) Create 5 lists composed of 4 random numbers between 1 and 100 and a string as final element.

Use random.randint(i,n) from numpy to get a list of n random numbers between 0 and i-1.

In [177]:
data = np.random.randint(5,30,size=10)
df = pd.DataFrame(data, columns=['F1'])

print(df)

# df = pd.DataFrame(np.random.randint(0,100,size=(100, 4)), columns=list('ABCD'))

   F1
0  25
1  22
2  16
3  10
4  26
5  19
6  24
7  14
8  25
9  10


#### &#x270d; 14.3) Create a dataframe named data, composed of these 5 lists. We will call the columns F1, F2, F3, F4 and STR.



In [191]:
from numpy.random import randn
data = pd.DataFrame(randn(5,5),columns='F1 F2 F3 F4 STR'.split())
data

F1        F2        F3        F4       STR
0 -0.779312  2.509836  0.928342 -0.909764 -0.488716
1 -0.397890  0.553348 -1.174959  1.534847  0.233664
2 -1.895684 -0.524575  0.049996 -1.166536 -0.160575
3 -0.898161  1.260797  0.007961  0.192886  0.224885
4  0.416375  0.376015 -0.906487  1.224524 -0.064446

#### &#x270d; 14.4) Print the STR feature of each element.

In [195]:
df[['F4','STR']]

KeyError: "None of [Index(['F4', 'STR'], dtype='object')] are in the [columns]"

#### &#x270d; 14.5) Create a new dataframe composed of the rows 1, 3 and 4 of data. Compute its F3 mean.



#### &#x270d; 14.6) Let's go back to the dataframe `data`. Create a new dataframe composed of the daqta rows where F1 > 40.

